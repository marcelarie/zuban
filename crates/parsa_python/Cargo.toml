[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
