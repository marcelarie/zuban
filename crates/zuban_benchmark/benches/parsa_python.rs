use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
