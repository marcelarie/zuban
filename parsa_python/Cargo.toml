[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
