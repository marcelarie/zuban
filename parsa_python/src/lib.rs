// Heavily inspired by https://docs.python.org/3.10/reference/grammar.html
// and adapted to Python 3.13

#![recursion_limit = "2048"]
mod tokenizer;

use parsa::{create_grammar, Grammar};
pub use parsa::{CodeIndex, NodeIndex};

pub use crate::tokenizer::TerminalType;
use crate::tokenizer::{PyTerminal, PythonTokenizer};

create_grammar!(
    static PYTHON_GRAMMAR, struct PythonGrammar, struct PyTree, struct PyNode,
    enum PyNodeType, enum NonterminalType, PythonTokenizer, PyTerminal, TerminalType,

    soft_keywords=[
        Name: "match" | "case" | "_"
    ]

    // STARTING RULES
    // ==============
    file: stmt* Endmarker

    // GENERAL STATEMENTS
    // ==================
    stmt: @error_recovery
          simple_stmts | Newline
        | if_stmt | while_stmt | for_stmt | try_stmt | with_stmt
        | function_def | class_def | decorated | async_stmt | match_stmt
    simple_stmts: simple_stmt (";" simple_stmt)* [";"] Newline
    // NOTE: assignment MUST precede expression, otherwise parsing a simple assignment
    // will throw a SyntaxError.
    simple_stmt:
        | assignment | star_expressions | del_stmt | pass_stmt
        | import_name | import_from | global_stmt | nonlocal_stmt | assert_stmt
        | break_stmt | continue_stmt | return_stmt | raise_stmt | yield_expr
    async_stmt: "async" (function_def | with_stmt | for_stmt)

    // SIMPLE STATEMENTS
    // =================
    assignment:
        | (star_targets "=" )+ (yield_expr | star_expressions)
        | single_target annotation ["=" (yield_expr | star_expressions)]
        | single_target augassign (yield_expr | star_expressions)

    augassign: ("+=" | "-=" | "*=" | "@=" | "/=" | "%=" | "&=" | "|=" | "^=" |
                "<<=" | ">>=" | "**=" | "//=")
    return_stmt: "return" [star_expressions]
    raise_stmt: "raise" [expression ["from" expression]]
    global_stmt: "global" ",".Name+
    nonlocal_stmt: "nonlocal" ",".Name+
    del_stmt: "del" del_targets
    assert_stmt: "assert" expression ["," expression]
    pass_stmt: "pass"
    break_stmt: "break"
    continue_stmt: "continue"

    // Import statements
    // -----------------
    import_name: "import" dotted_as_names
    // note below: the ("." | "...") is necessary because "..." is tokenized as ELLIPSIS
    import_from:
        | "from" ("." | "...")* dotted_name "import" import_from_targets
        | "from" ("." | "...")+ "import" import_from_targets
    import_from_targets: "*" | "(" ",".import_from_as_name+ ","? ")" | ",".import_from_as_name+
    import_from_as_name: Name "as" name_definition | name_definition
    dotted_as_names: ",".dotted_as_name+
    dotted_as_name: dotted_name "as" name_definition | name_definition ["." dotted_name]
    dotted_name: [dotted_name "."] Name

    // COMPOUND STATEMENTS
    // ===================

    // Common elements
    // ---------------

    block: simple_stmts | Newline Indent stmt+ Dedent
    decorators: decorator+
    decorator: "@" named_expression Newline
    decorated: decorators (class_def | function_def | async_function_def)

    // Class definitions
    // -----------------

    class_def: "class" name_definition ["(" [arguments] ")"] ":" block

    // Function definitions
    // --------------------

    async_function_def: "async" function_def
    function_def: "def" name_definition function_def_parameters return_annotation? ":" block
    return_annotation: "->" expression

    // Function parameters
    // -------------------

    function_def_parameters: "(" [parameters] ")"

    parameters:
        // no-default
        | ",".param_no_default+ ["," [star_etc]]
        // no-default slash no-default default
        | ",".param_no_default+ "," "/" ["," [
                ",".param_no_default+ [
                    "," [
                             ",".param_with_default+ ["," [star_etc]]
                             | [star_etc]
                        ]]
                | star_etc
            ]]
        // no-default slash default
        | ",".param_no_default+ "," "/" ["," [
                ",".param_with_default+ ["," [star_etc]]
                | star_etc
            ]]
        // no-default default
        | ",".param_no_default+ "," ",".param_with_default+ (
            ["," [star_etc]]
            // no-default default slash default
            | "," "/" ["," [
                ",".param_with_default+ ["," [star_etc]]
                | star_etc
            ]]
        )
        // default slash default
        | ",".param_with_default+ "," "/" ["," [
                ",".param_with_default+ ["," [star_etc]]
                | star_etc
            ]]
        // default
        | ",".param_with_default+ ["," [star_etc]]
        // just star args
        | star_etc
    star_etc:
        | starred_param ["," ",".param_maybe_default+] ["," [double_starred_param ","?]]
        | "*" "," ",".param_maybe_default+ ["," [double_starred_param ","?]]
        | double_starred_param [","]
    param_no_default: name_definition annotation? !"="
    param_with_default: name_definition annotation? "=" expression
    param_maybe_default: name_definition annotation? ["=" expression ]
    starred_param: "*" name_definition star_annotation?
    double_starred_param: "**" name_definition annotation?
    annotation: ":" expression
    star_annotation: ":" (star_expression | expression)

    // If statement
    // ------------

    if_stmt: "if" named_expression ":" block ("elif" named_expression ":" block)* else_block?
    else_block: "else" ":" block

    // While statement
    // ---------------

    while_stmt: "while" named_expression ":" block else_block?

    // For statement
    // -------------

    for_stmt: "for" star_targets "in" star_expressions ":" block else_block?

    // With statement
    // --------------

    with_stmt: "with" with_items ":" block
    with_items: "(" ",".with_item+ ","? ")" | ",".with_item+
    with_item: expression ["as" star_target]

    // Try statement
    // -------------

    try_stmt: "try" ":" block ((except_block+ | except_star_block+) else_block? finally_block? | finally_block)

    // Except statement
    // ----------------

    except_block: "except" [except_expression] ":" block
    except_star_block: "except" "*" except_expression ":" block
    except_expression: expression ["as" name_definition]
    finally_block: "finally" ":" block

    // Match statement
    // ---------------

    match_stmt: "match" subject_expr ":" Newline Indent case_block+ Dedent
    subject_expr:
        | star_named_expression "," star_named_expressions?
        | named_expression
    case_block: "case" patterns guard? ":" block
    guard: "if" named_expression

    patterns: open_sequence_pattern | pattern
    pattern: as_pattern | or_pattern
    as_pattern: or_pattern "as" pattern_capture_target
    or_pattern: "|".closed_pattern+
    closed_pattern:
        | literal_pattern | class_pattern | wildcard_pattern
        | group_pattern | sequence_pattern | mapping_pattern
        | pattern_capture_target | value_pattern

    literal_pattern:
        | complex_number | signed_number | strings | bytes
        | "None" | "True" | "False"
    complex_number: signed_number ("+"|"-") Number
    signed_number: "-"? Number

    pattern_capture_target: !"_" name_definition
    wildcard_pattern: "_"
    value_pattern: dotted_name

    group_pattern: "(" pattern ")"
    sequence_pattern:
        | "[" maybe_sequence_pattern? "]"
        | "(" open_sequence_pattern? ")"
    open_sequence_pattern: maybe_star_pattern "," maybe_sequence_pattern?
    maybe_sequence_pattern: ",".maybe_star_pattern+ ","?
    maybe_star_pattern: star_pattern | pattern
    star_pattern: "*" (pattern_capture_target | wildcard_pattern)

    mapping_pattern:
        | "{" double_star_pattern? "}"
        | "{" ",".key_value_pattern+ ["," double_star_pattern?] "}"
    key_value_pattern: (literal_pattern | value_pattern) ":" pattern
    double_star_pattern: "**" pattern_capture_target ","?

    class_pattern:
        | dotted_name "(" ")"
        | dotted_name "(" param_patterns ","? ")"
    param_patterns:
        | ",".(pattern !"=")+ [",".(keyword_pattern)+]
        | ",".(keyword_pattern)+
    keyword_pattern: Name "=" pattern

    // Type statement
    // ---------------
    //type_alias: "type" Name [type_params] "=" expression


    // Type parameter declaration
    // --------------------------
    /*
    type_params: "[" type_param_seq  "]"

    type_param_seq: ",".type_param+ [","]

    type_param:
        | Name [type_param_bound]
        | "*" Name ":" expression
        | "*" Name
        | "**" Name ":" expression
        | "**" Name

    type_param_bound: ":" expression
    */

    // EXPRESSIONS
    // -----------

    expressions: ",".expression+ [","]
    expression: ternary | lambda
    ternary:? disjunction ["if" disjunction "else" expression]

    yield_expr: "yield" [yield_from | star_expressions]
    yield_from: "from" expression

    star_expressions: ",".(expression|star_expression)+ [","]
    star_expression: "*" bitwise_or
    star_named_expressions: ",".star_named_expression+ [","]
    star_named_expression:? "*" disjunction | named_expression

    named_expression: name_definition ":=" expression | expression

    disjunction:? [disjunction "or"] conjunction
    conjunction:? [conjunction "and"] inversion
    inversion:? "not" inversion | comparison

    // Comparison operators
    // --------------------

    comparison:? bitwise_or (comp_op bitwise_or)*
    comp_op: "<"|">"|"=="|">="|"<="|"!="|"in"|"not" "in"|"is"|"is" "not"

    // Bitwise operators
    // -----------------

    bitwise_or:?   [bitwise_or "|"] bitwise_xor
    bitwise_xor:? [bitwise_xor "^"] bitwise_and
    bitwise_and:? [bitwise_and "&"] shift_expr
    shift_expr:?  [shift_expr ("<<"|">>")] sum

    // Arithmetic operators
    // --------------------

    sum:? sum ("+"|"-") term | term
    term:? [term ("*"|"@"|"/"|"%"|"//")] factor
    factor:? ("+"|"-"|"~") factor | power
    power:? await_primary ["**" factor]

    // Primary elements
    // ----------------

    await_primary:? ["await"] primary
    primary:?
          primary "." Name
        | primary "(" [arguments | comprehension] ")"
        | primary "[" slices "]"
        | atom
    slices:? ",".(slice | named_expression | starred_expression)+ [","]
    slice: expression? ":" expression? [":" expression?]
    atom:
          "(" [tuple_content | yield_expr | named_expression | comprehension] ")"
        | "[" [star_named_expressions | comprehension] "]"
        | "{" [dict_content | star_named_expressions | dict_comprehension | comprehension] "}"
        | Name | Number | strings | bytes | "..." | "None" | "True" | "False"

    // Lambda functions
    // ----------------

    lambda: "lambda" [lambda_parameters] ":" expression

    // Lambda params is basically a repetition of normal params without annotations
    lambda_parameters:
        // no-default
        | ",".lambda_param_no_default+ ["," [lambda_star_etc]]
        // no-default slash no-default default
        | ",".lambda_param_no_default+ "," "/" ["," [
                ",".lambda_param_no_default+ [
                    "," [
                             ",".lambda_param_with_default+ ["," [lambda_star_etc]]
                             | [lambda_star_etc]
                        ]]
                | lambda_star_etc
            ]]
        // no-default slash default
        | ",".lambda_param_no_default+ "," "/" ["," [
                ",".lambda_param_with_default+ ["," [lambda_star_etc]]
                | lambda_star_etc
            ]]
        // no-default default
        | ",".lambda_param_no_default+ "," ",".lambda_param_with_default+ (
            ["," [lambda_star_etc]]
            // no-default default slash default
            | "," "/" ["," [
                ",".lambda_param_with_default+ ["," [lambda_star_etc]]
                | lambda_star_etc
            ]]
        )
        // default slash default
        | ",".lambda_param_with_default+ "," "/" ["," [
                ",".lambda_param_with_default+ ["," [lambda_star_etc]]
                | lambda_star_etc
            ]]
        // default
        | ",".lambda_param_with_default+ ["," [lambda_star_etc]]
        // just star args
        | lambda_star_etc
    lambda_star_etc:
        | lambda_starred_param ["," ",".lambda_param_maybe_default+] ["," [lambda_double_starred_param ","?]]
        | "*" "," ",".lambda_param_maybe_default+ ["," [lambda_double_starred_param ","?]]
        | lambda_double_starred_param [","]
    lambda_param_no_default: name_definition !"="
    lambda_param_with_default: name_definition "=" expression
    lambda_param_maybe_default: name_definition ["=" expression ]
    lambda_starred_param: "*" name_definition
    lambda_double_starred_param: "**" name_definition

    // LITERALS
    // ========

    fstring: FStringStart fstring_content* FStringEnd
    fstring_content:? FStringString | fstring_expr
    fstring_conversion: "!" Name
    fstring_expr: "{" expressions ["="] [ fstring_conversion ] [ fstring_format_spec ] "}"
    fstring_format_spec: ":" fstring_content*

    strings: (String | fstring)+
    bytes: Bytes+

    tuple_content: star_named_expression "," [star_named_expressions]

    // Dicts
    // -----

    dict_content: ",".(dict_starred | dict_key_value)+ [","]
    dict_starred: "**" bitwise_or
    dict_key_value: expression ":" expression

    // Comprehensions & Generators
    // ---------------------------

    comprehension: named_expression for_if_clauses
    for_if_clauses: async_for_if_clause+
    async_for_if_clause:? ["async"] sync_for_if_clause
    sync_for_if_clause: "for" star_targets "in" disjunction comp_if*
    comp_if: "if" disjunction
    dict_comprehension: dict_key_value for_if_clauses

    // FUNCTION CALL ARGUMENTS
    // =======================

    arguments:
        | ",".(starred_expression | named_expression !"=")+ ["," kwargs?]
        | kwargs
    kwargs:
        | ",".(kwarg | starred_expression)+ ","?
        | ",".(kwarg | starred_expression)+ "," ",".(kwarg | double_starred_expression)+ ","?
        | ",".(kwarg | double_starred_expression)+ ","?
    starred_expression: "*" expression
    double_starred_expression: "**" expression
    kwarg: Name "=" expression

    // ASSIGNMENT TARGETS
    // ==================

    // Generic targets
    // ---------------

    star_targets: ",".star_target+ [","]
    star_target:? "*"? (t_primary | star_target_brackets | name_definition)
    star_target_brackets: "(" [star_targets] ")" | "[" [star_targets] "]"

    single_target: t_primary | name_definition | "(" single_target ")"

    t_primary:?
          (
              t_primary "." Name
            | t_primary "(" [arguments|comprehension] ")"
            | atom
        ) &("."|"["|"(")
        | t_primary "[" slices "]"
        | t_primary "." name_definition

    name_definition: Name

    // Targets for del statements
    // --------------------------

    del_targets: ",".(t_primary | del_t_atom)+ [","]
    del_t_atom:? name_definition | "(" [del_targets] ")" | "[" [del_targets] "]"

);

pub fn parse(code: Box<str>) -> PyTree {
    // TODO is this really the best way? Especially for refactoring?!
    let mut code = code.to_string();
    if !code.ends_with('\n') {
        code += "\n";
    }
    PYTHON_GRAMMAR.parse(code.into())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_grammar() {
        let tree = parse("{foo: 1}\n".into());
        let root_node = tree.root_node();
        assert_eq!(
            root_node.type_(),
            PyNodeType::Nonterminal(NonterminalType::file)
        );
    }
}
