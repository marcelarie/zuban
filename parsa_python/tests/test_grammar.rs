extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
