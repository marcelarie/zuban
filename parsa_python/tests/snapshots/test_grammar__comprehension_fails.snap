---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-52
 Error(stmt): 9-20
  Error(simple_stmts): 9-20
   Error(simple_stmt): 9-20
    Error(star_expressions): 9-20
     Error(expression): 9-20
      Error(ternary): 9-20
       Error(disjunction): 9-20
        Error(conjunction): 9-20
         Error(inversion): 9-20
          Error(comparison): 9-20
           Error(bitwise_or): 9-20
            Error(bitwise_xor): 9-20
             Error(bitwise_and): 9-20
              Error(shift_expr): 9-20
               Error(sum): 9-20
                Error(term): 9-20
                 Error(factor): 9-20
                  Error(power): 9-20
                   Error(await_primary): 9-20
                    Error(primary): 9-20
                     Error(atom): 9-20
                      Keyword: 9-10 "["
                      Error(comprehension): 10-20
                       named_expression: 10-11
                        expression: 10-11
                         atom: 10-11
                          Number: 10-11 "1"
                       Error(for_if_clauses): 12-20
                        Error(async_for_if_clause): 12-20
                         Error(sync_for_if_clause): 12-20
                          Keyword: 12-15 "for"
                          star_targets: 16-17
                           name_definition: 16-17
                            Name: 16-17 "a"
                          Keyword: 18-20 "in"
 Error(stmt): 21-23
  Error(simple_stmts): 21-23
   simple_stmt: 21-23
    star_expressions: 21-23
     star_expression: 21-23
      Keyword: 21-22 "*"
      atom: 22-23
       Name: 22-23 "b"
 Error(stmt): 23-24
  Error(Keyword): 23-24 "]"
 stmt: 24-25
  Newline: 24-25 "\n"
 Error(stmt): 25-43
  Error(simple_stmts): 25-43
   Error(simple_stmt): 25-43
    Error(star_expressions): 25-43
     Error(expression): 25-43
      Error(ternary): 25-43
       Error(disjunction): 25-43
        Error(conjunction): 25-43
         Error(inversion): 25-43
          Error(comparison): 25-43
           Error(bitwise_or): 25-43
            Error(bitwise_xor): 25-43
             Error(bitwise_and): 25-43
              Error(shift_expr): 25-43
               Error(sum): 25-43
                Error(term): 25-43
                 Error(factor): 25-43
                  Error(power): 25-43
                   Error(await_primary): 25-43
                    Error(primary): 25-43
                     Error(atom): 25-43
                      Keyword: 25-26 "["
                      comprehension: 26-43
                       named_expression: 26-27
                        expression: 26-27
                         atom: 26-27
                          Number: 26-27 "1"
                       for_if_clauses: 28-43
                        sync_for_if_clause: 28-43
                         Keyword: 28-31 "for"
                         star_targets: 32-33
                          name_definition: 32-33
                           Name: 32-33 "a"
                         Keyword: 34-36 "in"
                         atom: 37-38
                          Name: 37-38 "b"
                         comp_if: 39-43
                          Keyword: 39-41 "if"
                          atom: 42-43
                           Name: 42-43 "a"
 Error(stmt): 44-48
  Error(Keyword): 44-48 "else"
 Error(stmt): 49-50
  Error(simple_stmts): 49-50
   simple_stmt: 49-50
    star_expressions: 49-50
     expression: 49-50
      atom: 49-50
       Name: 49-50 "c"
 Error(stmt): 50-51
  Error(Keyword): 50-51 "]"
 stmt: 51-52
  Newline: 51-52 "\n"
 Endmarker: 52-52 ""

